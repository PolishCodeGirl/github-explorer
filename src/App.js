import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import styled from 'styled-components';
import Div from 'styled-kit/Div';

import Form from './components/Form';
import UserInformation from './components/UserInformation';
import InfoBox from './components/InfoBox';
import Loader from './components/Loader';

const mapStateToProps = ({ githubAccounts, githubAccountsLoading, searchedName, error }) => ({ 
  githubAccounts,
  githubAccountsLoading,
  searchedName,
  error
});

const propTypes = {
  githubAccounts: PropTypes.arrayOf(PropTypes.object).isRequired,
  githubAccountsLoading: PropTypes.bool.isRequired,
  searchedName: PropTypes.string.isRequired,
  error: PropTypes.bool.isRequired
}

function App({ githubAccounts, githubAccountsLoading, searchedName, error }) {
  return (
    <Div column itemsCenter width='100%'>
      {error && <InfoBox type="error" message='Something went wrong, try again!' />}
      
      <H1>
        <span>YND</span> GitHub repositories explorer
      </H1>

      <AppContainer column itemsCenter justifyCenter width='100%' maxWidth={500}>
        <Form />

        {
          githubAccountsLoading ? <Loader /> :
          searchedName ? githubAccounts?.length === 0 ? <InfoBox type="error" message={`Not users available under ${searchedName}`} /> :
              <>
                <Info>{`Showing users for "${searchedName}"`}</Info> 
                {githubAccounts?.map(account => (
                  <UserInformation userName={account.login} reposUrl={account.repos_url} key={account.id} />
                ))}
              </> : null
        }
      </AppContainer>
    </Div>
  );
}

App.propTypes = propTypes;

export default connect(mapStateToProps)(App);

const AppContainer = styled(Div)`
  font-size: 16px;
  margin: 10px;
  word-break: break-word;
`;

const H1 = styled.h1`
  span {
    color: #32d2d1;
    font-weight: bold;
  }
`;

const Info = styled.p`
  color: #747474;
  margin: 15px 0;
  align-self: flex-start;
`;
