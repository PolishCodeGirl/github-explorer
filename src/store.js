import { createStore, applyMiddleware, compose } from "redux";
import thunk from 'redux-thunk';
import githubUsers from './reducers/githubUsers';

const composeEnhancers =
  typeof window === 'object' &&
  window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ ?   
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({}) : compose;

const enhancer = composeEnhancers(applyMiddleware(thunk));

const store = createStore(githubUsers, enhancer);

export const { dispatch } = store;

export default store;