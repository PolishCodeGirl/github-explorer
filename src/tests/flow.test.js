import React from 'react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { render, fireEvent } from './utils';
import '@testing-library/jest-dom/extend-expect';
import store from '../store';
import App from '../App';

const server = setupServer(
  rest.get('https://api.github.com/search/users', (request, response, context) => {
    const pages = {
      1: [
        {
          id: 1,
          login: 'sara',
          repos_url: 'https://api.github.com/users/sara/repos',
        },
        {
          id: 2,
          login: 'SaraVieira',
          repos_url: 'https://api.github.com/users/SaraVieira/repos',
        },
        {
          id: 3,
          login: 'sara8086',
          repos_url: 'https://api.github.com/users/sara8086/repos',
        },
      ],
    };

    return response(context.json({ items: pages[request.url.searchParams.get('page')] }));
  }),
  rest.get('https://api.github.com/users/sara/repos', (_, response, context) =>
    response(
      context.json([
        {
          id: 1,
          name: 'CapitalOne-Two-Three',
          description: 'Educational finance webapp for kids',
          stargazers_count: 10,
          html_url: 'https://github.com/sara/CapitalOne-Two-Three',
        },
        {
          id: 2,
          name: 'Reop 2',
          description: 'Description reop 2',
          stargazers_count: 0,
          html_url: 'https://github.com/sara/repo-2',
        },
        {
          id: 3,
          name: 'Reop 3',
          description: 'Description reop 3',
          stargazers_count: 1,
          html_url: 'https://github.com/sara/repo-3',
        },
        {
          id: 4,
          name: 'Reop 4',
          description: 'Description reop 4',
          stargazers_count: 2,
          html_url: 'https://github.com/sara/repo-4',
        },
        {
          id: 5,
          name: 'Reop 5',
          description: 'Description reop 5',
          stargazers_count: 3,
          html_url: 'https://github.com/sara/repo-5',
        },
        {
          id: 6,
          name: 'Reop 6',
          description: 'Description reop 6',
          stargazers_count: 4,
          html_url: 'https://github.com/sara/repo-6',
        },
      ])
    )
  ),
  rest.get('https://api.github.com/users/SaraVieira/repos', (_, response, context) => response(context.json([])))
);

test('The flow works', async () => {
  // Start server
  server.listen();

  // Render the app
  const { container, getByText, findByText, getByPlaceholderText, getByTestId } = render(<App />);

  // Title should be visible from the start
  expect(getByText('GitHub repositories explorer')).toBeVisible();

  // Type "sara" into the text field
  fireEvent.change(getByPlaceholderText('Enter username'), { target: { value: 'sara' } });

  // Click the "Search" button
  fireEvent.click(getByText('Search'));
  expect(getByTestId('loader')).toBeVisible();

  // Info label should appear after we get 'getGithubUsers' API response
  expect(await findByText('Showing users for "sara"')).toBeVisible();
  expect(store.getState().searchedName).toEqual('sara');
  expect(store.getState().githubAccounts).toHaveLength(3);

  // These three users should be visible
  expect(getByText('sara')).toBeVisible();
  expect(getByText('SaraVieira')).toBeVisible();
  expect(getByText('sara8086')).toBeVisible();

  // Click on the first user
  fireEvent.click(getByText('sara'));
  expect(getByTestId('loader')).toBeVisible();

  // Repository info should appear after we get 'getUserRepos' API response
  expect(await findByText('CapitalOne-Two-Three')).toBeVisible();
  expect(await findByText('Reop 2')).toBeVisible();
  expect(await findByText('Reop 3')).toBeVisible();
  expect(await findByText('Reop 4')).toBeVisible();
  expect(await findByText('Reop 5')).toBeVisible();
  expect(store.getState().userRepos.sara).toHaveLength(6);

  // Click 'Show more' to display more user repos
  fireEvent.click(getByText('Show more'));
  expect(await findByText('Reop 6')).toBeVisible();

  // Click on the first user
  fireEvent.click(getByText('SaraVieira'));
  expect(getByTestId('loader')).toBeVisible();
  expect(await findByText(`SaraVieira doesn't have any repositories`)).toBeVisible();

  // Make sure the UI didn't change
  expect(container).toMatchSnapshot();

  // Stop server
  server.resetHandlers();
  server.close();
});
