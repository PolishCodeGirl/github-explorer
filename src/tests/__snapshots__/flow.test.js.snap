// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The flow works 1`] = `
<div>
  <div
    class="sc-1qr1lby-0 fgxJn"
  >
    <h1>
      <span
        class="sc-jrsJCI gnRkIr"
      >
        YND
      </span>
       GitHub repositories explorer
    </h1>
    <div
      class="sc-1qr1lby-0 sc-pNWxx iUrzEx eJTqIa"
      width="100%"
    >
      <form
        class="sc-1qr1lby-0 cWObwi"
        width="100%"
      >
        <input
          class="sc-bdnylx goodOi"
          placeholder="Enter username"
          required=""
          type="text"
          value="sara"
        />
        <button
          class="sc-gtssRu eSBAEQ"
          type="submit"
        >
          Search
        </button>
      </form>
      <p
        class="sc-kEqYlL faKJPi"
      >
        Showing users for "sara"
      </p>
      <div
        class="sc-1qr1lby-0 khdmUm"
        width="100%"
      >
        <div
          class="sc-1qr1lby-0 UserInformation___StyledDiv-sc-1v6v454-0 doMRkI bWBKbH"
        >
          <p>
            sara
          </p>
          <span
            class="sc-fujyUd chdwBd"
          />
        </div>
        <div
          class="sc-1qr1lby-0 bMIGKt"
        >
          <div
            class="sc-1qr1lby-0 sc-dlnjPT cVDpBh egYwOy"
          >
            <div
              class="sc-1qr1lby-0 bKUFbA"
            >
              <a
                class="sc-hKFyIo ekfeMC"
                href="https://github.com/sara/CapitalOne-Two-Three"
                rel="noopener noreferrer"
                target="_blank"
              >
                CapitalOne-Two-Three
              </a>
              <p
                class="sc-jSFkmK KwDyk"
              >
                Educational finance webapp for kids
              </p>
            </div>
            <p
              class="sc-eCApGN brKmTP"
            >
              10 ★
            </p>
          </div>
          <div
            class="sc-1qr1lby-0 sc-dlnjPT cVDpBh egYwOy"
          >
            <div
              class="sc-1qr1lby-0 bKUFbA"
            >
              <a
                class="sc-hKFyIo ekfeMC"
                href="https://github.com/sara/repo-2"
                rel="noopener noreferrer"
                target="_blank"
              >
                Reop 2
              </a>
              <p
                class="sc-jSFkmK KwDyk"
              >
                Description reop 2
              </p>
            </div>
            <p
              class="sc-eCApGN brKmTP"
            >
              0 ★
            </p>
          </div>
          <div
            class="sc-1qr1lby-0 sc-dlnjPT cVDpBh egYwOy"
          >
            <div
              class="sc-1qr1lby-0 bKUFbA"
            >
              <a
                class="sc-hKFyIo ekfeMC"
                href="https://github.com/sara/repo-3"
                rel="noopener noreferrer"
                target="_blank"
              >
                Reop 3
              </a>
              <p
                class="sc-jSFkmK KwDyk"
              >
                Description reop 3
              </p>
            </div>
            <p
              class="sc-eCApGN brKmTP"
            >
              1 ★
            </p>
          </div>
          <div
            class="sc-1qr1lby-0 sc-dlnjPT cVDpBh egYwOy"
          >
            <div
              class="sc-1qr1lby-0 bKUFbA"
            >
              <a
                class="sc-hKFyIo ekfeMC"
                href="https://github.com/sara/repo-4"
                rel="noopener noreferrer"
                target="_blank"
              >
                Reop 4
              </a>
              <p
                class="sc-jSFkmK KwDyk"
              >
                Description reop 4
              </p>
            </div>
            <p
              class="sc-eCApGN brKmTP"
            >
              2 ★
            </p>
          </div>
          <div
            class="sc-1qr1lby-0 sc-dlnjPT cVDpBh egYwOy"
          >
            <div
              class="sc-1qr1lby-0 bKUFbA"
            >
              <a
                class="sc-hKFyIo ekfeMC"
                href="https://github.com/sara/repo-5"
                rel="noopener noreferrer"
                target="_blank"
              >
                Reop 5
              </a>
              <p
                class="sc-jSFkmK KwDyk"
              >
                Description reop 5
              </p>
            </div>
            <p
              class="sc-eCApGN brKmTP"
            >
              3 ★
            </p>
          </div>
          <div
            class="sc-1qr1lby-0 sc-dlnjPT cVDpBh egYwOy"
          >
            <div
              class="sc-1qr1lby-0 bKUFbA"
            >
              <a
                class="sc-hKFyIo ekfeMC"
                href="https://github.com/sara/repo-6"
                rel="noopener noreferrer"
                target="_blank"
              >
                Reop 6
              </a>
              <p
                class="sc-jSFkmK KwDyk"
              >
                Description reop 6
              </p>
            </div>
            <p
              class="sc-eCApGN brKmTP"
            >
              4 ★
            </p>
          </div>
          <span
            class="UserInformation__ShowMore-sc-1v6v454-1 dYJzOO"
            disabled=""
          >
            Show more
          </span>
        </div>
      </div>
      <div
        class="sc-1qr1lby-0 khdmUm"
        width="100%"
      >
        <div
          class="sc-1qr1lby-0 UserInformation___StyledDiv-sc-1v6v454-0 doMRkI bWBKbH"
        >
          <p>
            SaraVieira
          </p>
          <span
            class="sc-fujyUd hvGBic"
          />
        </div>
      </div>
      <div
        class="sc-1qr1lby-0 khdmUm"
        width="100%"
      >
        <div
          class="sc-1qr1lby-0 UserInformation___StyledDiv-sc-1v6v454-0 doMRkI bWBKbH"
        >
          <p>
            sara8086
          </p>
          <span
            class="sc-fujyUd hvGBic"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
